test/gtest/ucp/test_ucp_proto_select.cc

```diff
@@ -0,0 +1,447 @@
+    {
+        uct_base_iface_t *base = ucs_derived_of(iface, uct_base_iface_t);
+        ucs_status_t status    =
+            mock_t::actual_call(&base->internal_ops->iface_estimate_perf, iface,
+                                perf_attr);
+        if (status == UCS_OK) {
+            mock_component *self = find(base->md->component);
+            auto it = self->m_perf_attrs_funcs.find(self->m_iface_names[base]);
+            if (it != self->m_perf_attrs_funcs.end()) {
+                (it->second)(*perf_attr);
+            }
+        }
+
+        return status;
+    }
+
+    static ucs_status_t
+    iface_query_mock(uct_iface_h iface, uct_iface_attr_t *iface_attr)
+    {
+        ucs_status_t status =
+            mock_t::actual_call(&iface->ops.iface_query, iface, iface_attr);
+        if (status == UCS_OK) {
+            uct_base_iface_t *base = ucs_derived_of(iface, uct_base_iface_t);
+            mock_component *self   = find(base->md->component);
+            auto it = self->m_iface_attrs_funcs.find(self->m_iface_names[base]);
+            if (it != self->m_iface_attrs_funcs.end()) {
+                (it->second)(*iface_attr);
+            }
+        }
+
+        return status;
+    }
+
+    uct_component_h                           m_component;
+    uint64_t                                  m_flags_orig;
+    std::map<uct_base_iface_t *, std::string> m_iface_names;
+    std::map<std::string, iface_attr_func_t>  m_iface_attrs_funcs;
+    std::map<std::string, perf_attr_func_t>   m_perf_attrs_funcs;
+
+    typedef std::map<uct_component_h,
+                     std::unique_ptr<mock_component>> component_map_t;
+    static component_map_t        g_component_map;
+    static ucs_sys_dev_distance_t g_memory_distance;
+};
+
+mock_component::component_map_t mock_component::g_component_map;
+ucs_sys_dev_distance_t mock_component::g_memory_distance;
+
+
+struct proto_select_data {
+    std::string range_start;
+    std::string range_end;
+    std::string desc;
+    std::string config;
+};
+
+typedef std::vector<proto_select_data> proto_select_data_vec_t;
+
+class test_ucp_proto_select : public ucp_test {
+public:
+    static void get_test_variants(std::vector<ucp_test_variant>& variants)
+    {
+        add_variant(variants, UCP_FEATURE_TAG | UCP_FEATURE_AM);
+    }
+
+    virtual void init() override
+    {
+        modify_config("PROTO_ENABLE", "y");
+        modify_config("MAX_RNDV_LANES", "1");
+    }
+
+    virtual void cleanup() override
+    {
+        ucp_test::cleanup();
+        mock_component::cleanup();
+    }
+
+    static void
+    check_ep_config(entity &e, const proto_select_data_vec_t &data,
+                    ucp_proto_select_key_t key)
+    {
+        ucp_ep_config_t *config = ucp_worker_ep_config(e.worker(), 0);
+        check_proto_select(e.worker(), config->proto_select, data, key);
+    }
+
+    static void
+    check_rkey_config(entity &e, const proto_select_data_vec_t &data,
```

Comment:
Never called?

Reply:
Yes, it's never called so far.
To me the goal of this commit is to provide meaningful infra for writing mock tests for protocol selection. Currently I have just one test that is just an example. I think API for checking rkey config would be useful in the future